### PR TITLE
docs: add text-to-sql dev note

### PR DESCRIPTION
**Pull request by dhruvnathawani**
## Summary
Add a dev note documenting the enterprise-grade text-to-SQL SDG pipeline used to generate training data for Nemotron's SQL capabilities across PostgreSQL, MySQL, and SQLite.

## What's in the post

- Motivation: The "real-world gap" between academic benchmarks (Spider >85%) and production SQL (Spider 2.0 Lite <50%) — dialect specificity, dirty data, distractor tables, industry-specific schemas, complexity gradients
- Pipeline walkthrough: Conditional samplers (60 industries, 700 topics, 90 SQL concepts) → three-stage LLM generation (natural language prompt → database context with dirty data + distractors → SQL with chain-of-thought reasoning) → quality waterfall (SQLFluff syntax validation + 4-dimension LLM judge scoring)
- ASCII pipeline diagram showing the 4-stage flow (Conditional Samplers → Three-Stage LLM Generation → Quality Waterfall → Output)
- Deep dive into SubcategorySamplerParams for two-level conditional sampling (industry→topic, complexity→sql_concept)
- Quality waterfall breakdown: 300k generated → ~180k after SQLFluff → 96.5k after judge filtering (68% total rejection)
- Rich metadata table (industry, topic, complexity, concept, dialect, instruction style, 4 judge scores) enabling precision filtering for downstream training
- Discussion of chain-of-thought reasoning traces teaching models to think like Data Engineers
- Results: 96.5k filtered records, 3 SQL dialects, 60 industries, 700 topics, 89 SQL concepts, 100% syntax-verified, all judge dimensions ≥ 3/4
- 7 key takeaways covering conditional sampling, three-stage generation, dirty data, distractor tables, hard validators, multi-dimension scoring and CoT reasoning

## Files changed

- docs/devnotes/posts/text-to-sql.md (new)

### CODE REVIEW

File: docs/devnotes/posts/text-to-sql.md [hunk 1]
```diff
@@ -0,0 +1,801 @@
...
+
+7. **Multi-dimension scoring enables targeted filtering.** A query that scores 4 on Relevance but 1 on Efficiency tells you the model understood the task but wrote a bad plan. You can filter differently depending on what you're training for.
+
+8. **Chain-of-thought teaches reasoning, not just syntax.** Including reasoning traces in the training data teaches models to decompose problems, handle edge cases, and verify logic --- acting as a Data Engineer rather than a translator.
+
+---
+
+## **Looking Ahead: The Code Sandbox**
+
+The current Quality Waterfall validates syntax and assesses quality (LLM judges), but it doesn't verify *semantic correctness* --- whether the query actually returns the right results. We are actively implementing Data Designer's Code Sandbox to close this gap. The sandbox would execute generated SQL against a ground-truth database and compare results, enabling:
```
Comment (mvansegbroeck) [inline docs/devnotes/posts/text-to-sql.md]: >> we are actively implementing

Can we provide a rough estimate here, e.g. expected Q3?

File: docs/devnotes/posts/text-to-sql.md [hunk 2]
```diff
@@ -0,0 +1,801 @@
...
+    result = data_designer.create(config, num_records=32000, dataset_name=f"text2sql_{dialect}")
+```
+
+</details>
+
+---
+
+## **A Team Effort**
+
+This dataset builds on the foundation laid during our time at [Gretel.ai](https://gretel.ai) (creators of the [#1 trending synthetic text-to-SQL dataset on Hugging Face](https://huggingface.co/datasets/gretelai/synthetic_text_to_sql)). Today, we're proud to bring that DNA into NVIDIA, building the data infrastructure that powers the next generation of Nemotron models.
```
Comment (mvansegbroeck) [inline docs/devnotes/posts/text-to-sql.md]: I'd cut the Gretel reference here. The work stands on its own and we want to keep the focus on the NVIDIA/Nemotron story.

File: docs/devnotes/posts/text-to-sql.md [hunk 3]
```diff
@@ -0,0 +1,801 @@
...
+    TEXT_TO_SQL_LLM_JUDGE_PROMPT_TEMPLATE as SQL_JUDGE_PROMPT,
+    JUDGE_NATURALNESS_PROMPT, DATA_QUALITY_JUDGE_PROMPT, KNOWLEDGE_DEP_JUDGE_PROMPT,
+    SQL_CONTEXT_PROMPT_SQLITE, SQL_CONTEXT_PROMPT_MYSQL, SQL_CONTEXT_PROMPT_POSTGRES,
+    SQL_PROMPT_SQLITE, SQL_PROMPT_MYSQL, SQL_PROMPT_POSTGRES,
+    SQL_CONTEXT_JUDGE_PROMPT_SQLITE, SQL_CONTEXT_JUDGE_PROMPT_MYSQL, SQL_CONTEXT_JUDGE_PROMPT_POSTGRES,
+)
+from rubrics import SQL_SCORES, PROMPT_SCORES, DATA_QUALITY_SCORES, KNOWLEDGE_DEP_SCORES
+
+METADATA_FILE = "text2sql_seed.json"
+MODEL_NAME = "Qwen/Qwen3-235B-A22B-Thinking-2507"
```
Comment (mvansegbroeck) [inline docs/devnotes/posts/text-to-sql.md]: should we use the same model name here as in the try-it-yourself sample (`nvidia-text`)? If it does not matter in the devnote, then would go for Nemotron model for now

File: docs/devnotes/posts/text-to-sql.md (hunk 1 above)
Comment (dhruvnathawani) [reply docs/devnotes/posts/text-to-sql.md]: I think we should not make promises on timelines

Better to just say next steps / improvements

File: docs/devnotes/posts/text-to-sql.md [hunk 4]
```diff
@@ -0,0 +1,801 @@
...
+))
+
+# Generate
+data_designer = DataDesigner()
+preview = data_designer.preview(config, num_records=10)
+preview.display_sample_record()
+```
+
+<details markdown>
+<summary><strong>Full source: <code>sdg_qwen_235b.py</code> (production pipeline)</strong></summary>
```
Comment (mvansegbroeck) [inline docs/devnotes/posts/text-to-sql.md]: Can we rename this file? "qwen" isn't not a key id for the filename, but NDD is. Something like `sdg_ndd_text2sql.py` or so?

File: docs/devnotes/posts/text-to-sql.md (hunk 1 above)
Comment (mvansegbroeck) [reply docs/devnotes/posts/text-to-sql.md]: Makes sense. In that case let's soften the statement to "A natural next step would be adding Code Sandbox support to Data Designer". This keeps it forward-looking without committing to a date or to use implementing this now.

File: docs/devnotes/posts/text-to-sql.md (hunk 3 above)
Comment (dhruvnathawani) [reply docs/devnotes/posts/text-to-sql.md]: I changed it to be a generic model name nvidia-text

File: docs/devnotes/posts/text-to-sql.md [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+
+Data Designer's `ValidationColumnConfig` with `CodeValidatorParams` runs generated SQL through SQLFluff, a dialect-aware SQL linter:
+
+```python
+config.add_column(dd.ValidationColumnConfig(
+    name="sql_validity",
+    validator_type=dd.ValidatorType.CODE,
+    target_columns=["sql"],
+    validator_params=dd.CodeValidatorParams(code_lang=dd.CodeLang.SQL_ANSI),
+    batch_size=10,
```
[the remark this answers is not in the document]
Comment (dhruvnathawani) [reply docs/devnotes/posts/text-to-sql.md]: Good point, fixed

File: docs/devnotes/posts/text-to-sql.md [hunk 6]
```diff
@@ -0,0 +1,810 @@
...
+
+- Describes a **business problem**, not a SQL specification (no SQL jargon allowed)
+- Matches the sampled instruction style, linguistic register, and politeness level
+- Implicitly requires the sampled SQL concept, task type, data quality handling, and knowledge dependency
+- Uses realistic thresholds appropriate for small sample data (5-10 rows per table)
+
+```python
+config.add_column(dd.LLMTextColumnConfig(
+    name="sql_prompt",
+    model_alias="prompt_gen",
```
Comment (nabinchha) [inline docs/devnotes/posts/text-to-sql.md:234]: It looks like a model config for this alias is missing in the `DataDesignerConfigBuilder(...)` construction above. Same comment for other model aliases...

Comment (nabinchha) [reply docs/devnotes/posts/text-to-sql.md:234]: perhaps these snippets are not supposed to be runnable? either way, would be good to clarify.

File: docs/devnotes/posts/text-to-sql.md [hunk 7]
```diff
@@ -0,0 +1,810 @@
...
+
+> *"The SQL should only JOIN or reference tables that are strictly necessary to answer the prompt. The database context may include distractor tables that look relevant but are not needed -- penalize queries that unnecessarily join or reference these tables."*
+
+Each judge provides a score *and* reasoning for each dimension, making it easy to diagnose why a record scored low. Expression columns extract numeric scores into flat columns for downstream filtering:
+
+```python
+config.add_column(dd.ExpressionColumnConfig(
+    name="sql_relevance_score",
+    expr="{{ sql_judge_result.relevance.score if sql_judge_result.relevance.score is not none else '' }}",
+))
```
Comment (nabinchha) [inline docs/devnotes/posts/text-to-sql.md:379]: are we intentionally leaving out LLMJudgeColumn config out of here and jumping straight to score extraction?

File: docs/devnotes/posts/text-to-sql.md [hunk 8]
```diff
@@ -0,0 +1,810 @@
...
+
+# Prompt templates and judge rubrics (see prompts.py and rubrics.py)
+from prompts import (
+    SQL_PROMPT_SYSTEM_PROMPT, SQL_PROMPT_PROMPT,
+    SQL_CONTEXT_SYSTEM_PROMPT, SQL_SYSTEM_PROMPT,
+    TEXT_TO_SQL_LLM_JUDGE_PROMPT_TEMPLATE as SQL_JUDGE_PROMPT,
+    JUDGE_NATURALNESS_PROMPT, DATA_QUALITY_JUDGE_PROMPT, KNOWLEDGE_DEP_JUDGE_PROMPT,
+    SQL_CONTEXT_PROMPT_SQLITE, SQL_CONTEXT_PROMPT_MYSQL, SQL_CONTEXT_PROMPT_POSTGRES,
+    SQL_PROMPT_SQLITE, SQL_PROMPT_MYSQL, SQL_PROMPT_POSTGRES,
+    SQL_CONTEXT_JUDGE_PROMPT_SQLITE, SQL_CONTEXT_JUDGE_PROMPT_MYSQL, SQL_CONTEXT_JUDGE_PROMPT_POSTGRES,
```
Comment (nabinchha) [inline docs/devnotes/posts/text-to-sql.md]: how will these be accessible?

File: docs/devnotes/posts/text-to-sql.md (hunk 6 above)
Comment (dhruvnathawani) [reply docs/devnotes/posts/text-to-sql.md:234]: Yes this is a snippet of code, not supposed to be fully runnable, we have recipes for that, maybe we can link to those

File: docs/devnotes/posts/text-to-sql.md (hunk 8 above)
Comment (dhruvnathawani) [reply docs/devnotes/posts/text-to-sql.md]: Good point, going to have a self contained version here and link to recipe for the full version

File: docs/devnotes/posts/text-to-sql.md (hunk 7 above)
Comment (dhruvnathawani) [reply docs/devnotes/posts/text-to-sql.md:379]: Addressed """Each judge provides a score *and* reasoning for each dimension, making it easy to diagnose why a record scored low. After configuring the five `LLMJudgeColumnConfig` columns (see the [full recipe](../../recipes/code_generation/enterprise_text_to_sql/) for complete judge definitions), expression columns extract numeric scores into flat columns for downstream filtering:"""

File: docs/devnotes/posts/text-to-sql.md [hunk 9]
```diff
@@ -0,0 +1,598 @@
...
+| **Nemotron Super (after synthetic text-to-SQL SFT data)** | **41.80** |
+
+**Caveat on BIRD:** BIRD measures *execution accuracy* (EX) --- whether the query returns the correct result set when run against the ground-truth database. This is stricter than exact-match or string similarity, but it can also be inflated by semantically different queries that happen to produce identical result sets on small test data. BIRD's dev set includes dirty data, external knowledge requirements, and multi-table schemas, making it more representative of production SQL than earlier benchmarks like Spider --- but it does not cover all production challenges (e.g., multi-statement transactions, DDL, stored procedures, or the hundreds-of-tables schemas common in enterprise warehouses). Results here are on the 1,534-sample dev split averaged over 5 runs.
+
+---
+
+## **Key Takeaways**
+
+1. **Conditional sampling prevents incoherent records.** `SubcategorySamplerParams` ensures "Geospatial SQL" only appears with "Advanced" complexity, and "Electronic Health Records" only appears with "Healthcare". Independent samplers would produce nonsensical combinations that confuse training.
+
```
[the remark this answers is not in the document]
Comment (dhruvnathawani) [reply docs/devnotes/posts/text-to-sql.md:446]: This is too nit

**Review by mvansegbroeck — APPROVED** [review]

**Comment by andreatgretel** [comment]
> `mkdocs.yml:75-85`

the new dev note isn't listed in the `nav` section of mkdocs.yml - all the other dev notes have entries there (lines 78-85). it'll still show up on the blog index via auto-discovery, but it won't appear in the sidebar. should go after "Async All the Way Down" since it's dated 2026-03-11.

File: docs/devnotes/posts/text-to-sql.md [hunk 10]
```diff
@@ -0,0 +1,598 @@
...
+))
+
+config.add_column(dd.SamplerColumnConfig(
+    name="sql_concept",
+    sampler_type=dd.SamplerType.SUBCATEGORY,
+    params=dd.SubcategorySamplerParams(
+        category="sql_complexity",
+        values={
+            "Beginner":     ["Basic SELECT Statements", "WHERE Clauses", "Simple Aggregations", ...],
+            "Intermediate": ["Window Functions", "Recursive CTEs", "Correlated Subqueries", ...],
```
Comment (andreatgretel) [inline docs/devnotes/posts/text-to-sql.md]: "Recursive CTEs" is listed under Intermediate here, but the recipe included lower on the page via `--8<--` (line 240 of `enterprise_text_to_sql.py`) puts it under Advanced. your own intro also frames recursive CTEs as a senior-engineer concept. maybe swap it with "CASE Expressions" to match the recipe?

**Comment by andreatgretel** [comment]
> `docs/assets/recipes/code_generation/enterprise_text_to_sql.py:546`
>
> ```python
> expr=f"{{{{ {judge_name}.{rubric}.score if {judge_name}.{rubric}.score else ' ' }}}}"
> ```

this still uses a truthy check for score extraction, but the inline snippet in the dev note (line 378) was fixed to use `is not none`. since both appear on the same rendered page via the `--8<--` include, readers see contradictory patterns. the truthy version also silently drops legitimate score-0 values. Claude Code caught this one.

File: docs/devnotes/posts/text-to-sql.md [hunk 11]
```diff
@@ -0,0 +1,598 @@
...
+
+2. **Three-stage generation beats one-shot.** Separating prompt, schema, and query generation ensures the SQL actually references the tables that exist. One-shot generation frequently hallucinates tables.
+
+3. **Dirty data must be intentional.** Explicitly prompting for anti-patterns (dates as text, currency symbols, JSON blobs) forces the model to learn defensive SQL. Clean schemas produce clean-only training data.
+
+4. **Distractor tables teach schema linking.** Injecting semantically similar but irrelevant tables forces the model to *read* the schema instead of guessing from table names. This is the skill gap between academic benchmarks and production.
+
+5. **Per-dialect generation avoids lowest-common-denominator SQL.** Rather than generating ANSI SQL and hoping it works everywhere, the pipeline produces dialect-specific schemas and queries with appropriate syntax (`strftime` vs `DATE_SUB` vs `interval`, `REPLACE()` vs `regexp_replace`). Each dialect gets its own tailored prompts, validators, and judge prompts.
```
Comment (andreatgretel) [inline docs/devnotes/posts/text-to-sql.md]: nit: `REPLACE()` is available in all three dialects (SQLite, MySQL, PostgreSQL) - it's not really a dialect-specific function. the `strftime` vs `DATE_SUB` vs `interval` example right before it already makes the point well. maybe just drop the `REPLACE()` vs `regexp_replace` part? this was flagged in a prior review too.

**Comment by andreatgretel** [comment]
ran [slop-guard](https://github.com/eric-tramel/slop-guard) on the post - scored 23/100, which is in the same range as some existing dev notes (structured-outputs is 29, design-principles is 21), so not blocking. a couple easy wins if you want to tighten the prose: "The key insight:" on line 33 could just be cut (state the insight directly), "performance drops significantly" on line 23 could use the actual number instead, and the "X, not Y" contrast pattern shows up 5 times ("feature, not a bug", "reasoning, not just syntax", "hours, not months", etc.) - effective in moderation but gets repetitive across the post.

File: docs/devnotes/posts/text-to-sql.md [hunk 12]
```diff
@@ -0,0 +1,598 @@
...
+
+Rather than relying on LLM creativity alone for diversity, the pipeline samples structured metadata that deterministically controls every axis of variation. A JSON taxonomy file defines the problem space:
+
+| Axis | Categories | Subcategories | Role |
+|------|-----------|---------------|------|
+| Industry sector | 60 | ~700 topics | Domain grounding (Healthcare, FinServ, Gaming, ...) |
+| SQL complexity | 3 tiers | 89 concepts | Difficulty level (Beginner → Advanced) |
+| SQL task type | 12 categories | 94 concepts | What the query does (analytics, transformation, ...) |
+| Data quality | 5 challenges | 12 concepts | Dirty data to inject and clean |
+| Knowledge dependency | 3 categories | 8 concepts | Implicit reasoning required |
```
Comment (andreatgretel) [inline docs/devnotes/posts/text-to-sql.md]: the taxonomy table says "3 categories | 8 concepts" for knowledge dependency, but the recipe included on this page defines 3 categories x 3 subcategories = 9 concepts. small thing, but the demo recipe shouldn't exceed the stated total.

File: docs/devnotes/posts/text-to-sql.md (hunk 10 above)
Comment (dhruvnathawani) [reply docs/devnotes/posts/text-to-sql.md]: Good catch, changed to match recipe

File: docs/devnotes/posts/text-to-sql.md (hunk 11 above)
Comment (dhruvnathawani) [reply docs/devnotes/posts/text-to-sql.md]: Good point, changed

File: docs/devnotes/posts/text-to-sql.md (hunk 12 above)
Comment (dhruvnathawani) [reply docs/devnotes/posts/text-to-sql.md]: I specifically remember putting 9 LOL, thanks!
